Configurable PCI BDF Device IDs

scholzp opened the pull request:
Related to https://github.com/cobaltcore-dev/cobaltcore/issues/287.

# Configurable BDF Design

## Background
Configurable PCI BDF (Bus, Device, Function) allows for choosing the address that a guest can see a PCI device. Qemu, for example, implements the `addr` option for the `device` argument. The `addr` argument allows defining the device and function part of a device's BDF in the form `addr=DD.F`, where `D` denotes a device ID in the range of [0..32] and `F` a function ID in a range of [0..7]. Qemu denotes the bus number with the option `busnr`, where a valid expression is `busnr=pci.2` to denote the second PCI bus, for example. [0]

With a series of pull requests, we try to mimic the command-line interface of Qemu and introduce new command line parameters to CHV, where applicable to similarly achieve configurable BDFs.

An example of a BDF for Bus=1, Device=3, and Function=5 is `1:03.5`. [1] 

## Current status
CHV allows defining multiple PCI domains via the `pci_segments` option. For all PCI segments, CHV creates a dedicated host bridge. Via the option argument `pci_segment`, one can define to which domain a device is added.

Other than that, CHV currently has no support to influence what exact BDF is assigned to a device, other than the order in which devices are specified in the VM configuration. This is done by calling the `next_device_bdf` function of the
struct `PciSegment` for each PCI device to create. `next_device_bdf` returns the a BDF with the following free device ID and a fixed function ID of 0. `PciSegment` queries the underlying `PciBus` to obtain the following device ID, which we also must modify.

Currently, there is no support for multi-function devices in CHV.

## Proposed changes
We split the configurable BDF feature into separate PRs with the goals defined in The following subsections.

The BAR address allocation is able to handle the changes listed below. So I don't expect the need for changes there.

### 1. Definition of Device IDs
To define a device ID, we first must implement some notion for specifying them via the API or command line. For this, we introduce the `addr` option argument, similar to Qemu. The expected format is `addr=DD` with `DD` being in the range
of [0..32]. This argument is stored in a device's configuration and must be propagated to `next_device_bdf`. This function will be replaced by `specific_device_bdf` to allow for specifying a concrete device identifier. The steps for implementing configurable device IDs are as follows: 
- [x] Introduce function `allocate_device_id` to struct `PciBus` to reserve a device ID, depending on the argument given. If no device ID is specified, use the current mechanism to obtain an ID. This replaces `next_device_id`. 
- [x] Introduce function `allocate_device_bdf` to struct `PciSegment` to allocate a BDF depending on function parameters. If no device parameter is specified, fall back to the algorithm currently used by `next_device_bdf`. Replaces `next_device_bdf`.
- [x] Introduce `addr` parameter to all forms of VirtIo device configurations
 and adjust parsers
- [x] Add unit tests covering `allocate_device_id` in `PciBus`
- [x] Add unit tests covering `allocate_device_bdf` in `PciSegment`
- [x] Add unit tests covering the new argument and its parsing

## References
[0] https://qemu-project.gitlab.io/qemu/system/device-emulation.html#device-buses

[1] https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation

phip1611 commented on pci/src/bus.rs:
```diff
@@ -180,6 +180,30 @@ impl PciBus {
+    pub fn next_device_id_with_hint(&mut self, hint: &Option<u16>) -> Result<u32> {
+        if let Some(id) = hint {
+            let id = *id;
+            if (id as usize) >= NUM_DEVICE_IDS {
```
not sure but it might could be `>`

phip1611 commented on pci/src/bus.rs:
```diff
@@ -180,6 +180,30 @@ impl PciBus {
         Err(PciRootError::NoPciDeviceSlotAvailable)
     }
 
+    pub fn next_device_id_with_hint(&mut self, hint: &Option<u16>) -> Result<u32> {
```
please add unit tests for this function

phip1611 commented on vmm/src/device_manager.rs:
```diff
@@ -895,6 +895,7 @@ struct MetaVirtioDevice {
     iommu: bool,
     id: String,
     pci_segment: u16,
+    pci_device_hint: Option<u16>,
```
would `pci_bdf_hint` or `pci_addr_hint` be a better name? Or can we drop the "hint" at all? It is not an hint. We either get the requested PCI address or we fail

phip1611 commented on vmm/src/vm_config.rs:
```diff
@@ -276,6 +276,10 @@ pub struct DiskConfig {
+    // #[serde(default)]
+    // pub bus_hint: Option<u16>,
+    #[serde(default)]
+    pub addr_hint: Option<u16>,
```
same here, not sure of hint is the right name.

phip1611 reviewed (changes requested):
LGTM on a first glance. This however must be done at least also for virtio-net devices. So far I only see this is done for virtio-blk

phip1611 commented on pci/src/bus.rs:
```diff
@@ -166,15 +166,42 @@ impl PciBus {
+    /// Allocates a PCI device ID on the bus.
+    ///
+    /// - `id`: ID to allocate on the bus. If [`None`], the next free
+    ///         device ID on the bus is allocated, else the ID given is
```
This will not pass cargo fmt (or cargo clippy?). `device ID` should start on the same level as the \` from the line above

phip1611 commented on pci/src/bus.rs:
```diff
@@ -484,3 +511,98 @@ fn parse_io_config_address(config_address: u32) -> (usize, usize, usize, usize)
+
+#[cfg(test)]
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
```
is this comment from ChatGPT? :D

phip1611 commented on pci/src/bus.rs:
```diff
@@ -484,3 +511,98 @@ fn parse_io_config_address(config_address: u32) -> (usize, usize, usize, usize)
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
+    use super::*;
+
+    mod pci_bus_tests {
```
you are already in `pci::bus::tests`, no need for another nesting level

phip1611 commented on pci/src/bus.rs:
```diff
@@ -484,3 +511,98 @@ fn parse_io_config_address(config_address: u32) -> (usize, usize, usize, usize)
+
+        #[test]
+        // Test that requesting specific ID work
+        fn allocate_device_id_request_id() -> std::result::Result<(), Box<dyn std::error::Error>> {
```
please return `result::Result<(), Box<dyn error::Error>>` or even `Result<(), Box<dyn Error>>`

phip1611 commented on pci/src/bus.rs:
```diff
@@ -169,8 +169,8 @@ impl PciBus {
     /// - `id`: ID to allocate on the bus. If [`None`], the next free
-    ///         device ID on the bus is allocated, else the ID given is
-    ///         allocated
+    ///   device ID on the bus is allocated, else the ID given is
```
this changes should be merged into the commit where you introduced these lines, please

phip1611 commented on pci/src/bus.rs:
```diff
@@ -191,7 +191,7 @@ impl PciBus {
                 }
             } else {
-                return Err(PciRootError::InvalidPciDeviceSlot(id as usize));
+                Err(PciRootError::InvalidPciDeviceSlot(id as usize))
```
same here

phip1611 commented on vmm/src/pci_segment.rs:
```diff
@@ -163,15 +163,22 @@ impl PciSegment {
     }
 
-    pub(crate) fn next_device_bdf(&self) -> DeviceManagerResult<PciBdf> {
+    /// Allocates a device BDF on this PCI segment
```
- missing final period

phip1611 commented on vmm/src/pci_segment.rs:
```diff
@@ -480,3 +539,99 @@ impl Aml for PciSegment {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
+    use super::*;
+
+    mod pci_segment_tests {
```
same here, unnencessary level of nesting

phip1611 commented on vmm/src/pci_segment.rs:
```diff
@@ -480,3 +539,99 @@ impl Aml for PciSegment {
+
+#[cfg(test)]
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
```
again this comment; please remove

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1068,19 @@ impl RateLimiterGroupConfig {
     }
 }
 
+fn parse_addr(s: &str) -> std::result::Result<(u8, u8), OptionParserError> {
```
Please return `result::Result` or just `Result<(u8, u8)>`. Further, please add inline comments, like this: `Result<(u8 /* device */, u8 /* function */)>`

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1077,7 +1090,7 @@ impl DiskConfig {
          id=<device_id>,pci_segment=<segment_id>,rate_limit_group=<group_id>,\
          queue_affinity=<list_of_queue_indices_with_their_associated_cpuset>,\
-         serial=<serial_number>";
+         serial=<serial_number>, addr=<DD.F>";
```
- there should be no additional space (as for the other params)
- I think `DD.F` doesn't really help that much. What about `addr=<DD.F, e.g. 13.0>`? But I am also not sure. Maybe your solution is the best. hm.

scholzp replied:
What about `addr=<two_digit_pci_device_id.one_digit_pci_function_id>`?

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1302,7 +1324,7 @@ impl NetConfig {
     bw_size=<bytes>,bw_one_time_burst=<bytes>,bw_refill_time=<ms>,\
     ops_size=<io_ops>,ops_one_time_burst=<io_ops>,ops_refill_time=<ms>,pci_segment=<segment_id>\
-    offload_tso=on|off,offload_ufo=on|off,offload_csum=on|off\"";
+    offload_tso=on|off,offload_ufo=on|off,offload_csum=on|off, addr=DD.F\"";
```
unnecessary space!

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1078,7 +1078,15 @@ fn parse_addr(s: &str) -> std::result::Result<(u8, u8), OptionParserError> {
         .parse::<u8>()
         .map_err(|_| OptionParserError::InvalidSyntax(s.to_owned()))?;
-    Ok((a, b))
+    let f = if b != 0 {
```
This is hard to read.

Please add a check on a dedicated line and `return Err` in the corresponding `if cond {}`. After that, please return `Ok(b)`

phip1611 reviewed (changes requested):
LGTM for the workshop! For the purpose of our workshop (if it works), please leave it as is and focus on libvirt and libvirt-tests. 

Before we merge this, we need to address a few minor things.

scholzp committed:
vmm: Use device ID part of bdf for hotplugging

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

phip1611 commented on vmm/src/vm_config.rs:
```diff
@@ -479,6 +488,7 @@ pub struct PmemConfig {
     pub id: Option<String>,
     #[serde(default)]
     pub pci_segment: u16,
+    pub bdf_device_id: Option<u8>,
```
I wonder whether `bdf_device_id` is a bit "doppeltgemoppelt" and `bdf_device` is sufficient in conjunction with `bdf_function`?

scholzp replied:
I'm not sure. `bdf_device` was the first version I had in mind but thought that could might be a confused with a device's BDF. In conjunction with `bdf_function` this should work out, but that will be a feature that has not a high priority right now, as we don't support multi-function devices. Nevertheless, I tend to change this to `bdf_device`, If you agree.

scholzp committed:
vmm: Write directly to member  when creating new devices

Virtio PCI devices are created in a set of nested functions. In each
of this function a vector is created to add created devices to, only
to be appended to the vector of the higher nesting level. Those nested
vectors are unnecessary as we can directly write to the member of.

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

scholzp commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1068,26 @@ impl RateLimiterGroupConfig {
+        .split_once('.')
+        .ok_or(OptionParserError::InvalidSyntax(s.to_owned()))?;
+    let device_id = device_str
+        .parse::<u8>()
```
Can we parse hex or only dez?

scholzp replied:
I added the ability to either enter the addresses in hex or decimal numbers.

phip1611 commented on pci/src/bus.rs:
```diff
@@ -484,3 +511,94 @@ fn parse_io_config_address(config_address: u32) -> (usize, usize, usize, usize)
 }
+
+#[cfg(test)]
+mod tests {
```
nit. by convention the module should be named `unit_tests`

scholzp replied:
I renamed both modules I added. That in `bus.rs` and in `pci_segment.rs`

phip1611 commented on pci/src/bus.rs:
```diff
@@ -484,3 +511,94 @@ fn parse_io_config_address(config_address: u32) -> (usize, usize, usize, usize)
+    use super::*;
+
+    #[derive(Debug)]
+    struct MocRelocDevice;
```
please add a unit test where you specify devices with explicit addresses 1, 3, 5 and check that other devices are implicitley added at addresses 2 and 4

scholzp replied:
I added the test `allocate_device_id_fills_gaps`

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1068,28 @@ impl RateLimiterGroupConfig {
+        .ok_or(OptionParserError::InvalidSyntax(s.to_owned()))?;
+    let device_id = device_str
+        .parse::<u8>()
+        .or_else(|_| u8::from_str_radix(device_str, 16))
```
You've build a footgun here :)

Imagine someone changes `a0` to `90` - suddenly total different meaning. You could try to look for a `0x` prefix - I think the debugcon console parsing code also does this

scholzp replied:
After checking Qemu's behavior, we decided to go with accepting hex only.

phip1611 commented on vmm/src/config.rs:
```diff
@@ -361,6 +361,15 @@ pub enum ValidationError {
     #[error("Payload configuration is not bootable")]
     PayloadError(#[from] PayloadConfigError),
+    /// Invalid PCI function ID string
+    #[error("Function ID not zero: {0}")]
```
I don't think this is an error. Currently, I'd prefer to just panic with a `todo!(device ID currently not supported)`.

And you are going to introduce function ID support in a follow-up..right? 

Or am I totally missing something and we never want to specify the function ID?

scholzp replied:
That's correct, although I don't know when the followup will come as we currently didn't specify a concrete time frame.
Function IDs do only make sense if we implement multifunction devices. I think support for mutlifunction devices is planned but not a priority right now.

phip1611 replied:
ok. Could you rephrase the error variant and the message to `Multi-function devices currently not supported`?

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1077,24 @@ impl RateLimiterGroupConfig {
+        .ok_or(ValidationError::PciBdfSyntax(s.to_owned()))?;
+
+    let device_id = u8::from_str_radix(device_str.strip_prefix("0x").unwrap_or(device_str), 16)
+        .map_err(|_| ValidationError::PciBdfHex(s.to_owned()))?;
```
I'd prefer to keep the inner error.


```suggestion
        .map_err(|e| ValidationError::PciBdfHex(e, s.to_owned()))?;
```

than you could also have a `#[source] ParseErrorType` in the enum variant definition

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1077,24 @@ impl RateLimiterGroupConfig {
+        .split_once('.')
+        .ok_or(ValidationError::PciBdfSyntax(s.to_owned()))?;
+
+    let device_id = u8::from_str_radix(device_str.strip_prefix("0x").unwrap_or(device_str), 16)
```
```suggestion
    let device_id_str = device_str.strip_prefix("0x").unwrap_or(device_str);
    let device_id = u8::from_str_radix(device_id_str, 16)
```

phip1611 commented on vmm/src/config.rs:
```diff
@@ -1068,6 +1077,24 @@ impl RateLimiterGroupConfig {
+    let function_id = u8::from_str_radix(function_str.strip_prefix("0x").unwrap_or(device_str), 16)
+        .map_err(|_| ValidationError::PciBdfHex(s.to_owned()))?;
+
+    if function_id != 0 {
```
please add a _why_ comment here. Why is this not allowed currently?

scholzp committed:
vmm: Add tests for BDF device address parsing in configs

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

scholzp committed:
vmm: Use `VecDeque` to collect virtio devices

We use `VecDeque` to sort devices implicitly. Devices whose config
contains a fixed BDF are added to the front, while those without a BDF
given are added to the back. Processing the `VecDeque` sequentially
from first to last then ensures that no clashes occur when assigning
BDFs to devices. Otherwise, we could end up in the case that we assigned
a BDF required by one device's config to one without a BDF.

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

hertrste commented on vmm/src/pci_segment.rs:
```diff
@@ -480,3 +539,96 @@ impl Aml for PciSegment {
+    }
+
+    #[test]
+    // Test the default bdf for an an segment with an empty bus (except for the root device)
```
nit: `for an an segment`

hertrste commented on vmm/src/vm_config.rs:
```diff
@@ -276,6 +276,10 @@ pub struct DiskConfig {
     #[serde(default)]
     pub queue_affinity: Option<Vec<VirtQueueAffinity>>,
+    // #[serde(default)]
+    // pub bus_hint: Option<u16>,
```
Not sure if this goes away in somer other commit, but do we want to have commented out code here?

scholzp replied:
No, we don't. At least not these two lines. :D

hertrste commented on vmm/src/config.rs:
```diff
@@ -1877,12 +1965,20 @@ impl ConsoleConfig {
+            (Some(a), Some(b))
+        } else {
+            (None, None)
+        };
```
It feels like we repeat this snippet of code quite some time. Can we extract that into some function or is there a reason not to do so?

scholzp replied:
I moved the logic to the `OptionParser` :)

hertrste reviewed (approved):
Very nice :)

scholzp committed:
pci: Allow for device ID allocation on a bus

Allocating a device ID is crucial for assigning a specific ID to a
device. We need this to implement configurable PCI BDF.

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

amphi commented on src/main.rs:
```diff
@@ -1211,6 +1214,24 @@ mod unit_tests {
+                    "--kernel",
+                    "/path/to/kernel",
+                    "--disk",
+                    "path=/path/to/disk/1,addr=15.0",
+                    "path=/path/to/disk/2",
+                ],
+                r#"{
+                    "payload": {"kernel": "/path/to/kernel"},
+                    "disks": [
+                        {"path": "/path/to/disk/1", "bdf_device": 21},
```
I think it is a bit weird that on the command line the bdf is a hex value, and using the REST API its a decimal value. Do we want to keep it like that?

I have no strong opinion on that, I just want to know what the others think?

phip1611 replied:
Good point! The CLI is consistent with QEMU. I personally tihnk this difference between CLI and API is not a big problem.

@alyssais (sitting next to me) also just told me that this seems okay

scholzp replied:
As noted before: Really good point!

Regarding this, we have one more inconsistency: The response message after adding a device via `add-disk` or `add-net` will contain JSON with `PciDeviceInfo`. `PciDeviceInfo` contains the BDF accessible with key `bdf` as a string in the format `Domain:Bus:Device:Function`. If we would like to have full consistency, then we would need to also add the BDF as a whole in the same format via CLI. We decided to go with a CLI interface similar to Qemu, so we add the `Device:Function` part via the `addr` option. In CHV, you can already specify the PCI domain on the CLI and API via the `segment` option. So we only continue the inconsistency, if you want so (which should not be an excuse to not improve the situation).

phip1611 replied:
when you upstream this, just note this down and create an issue describing the problem and outline a possible solution.

scholzp replied:
Will discuss this upstream

scholzp committed:
vmm: Add tests for `allocate_device_bdf` in `PciSegment`

Next to tests for `allocate_device_bdf`, we introduce a new constructor
`new_without_address_manager`, only available in the test build. As
there is no way to instantiate an `AddressManager` in the tests, we use
this constructor to work around this.

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

scholzp committed:
vmm: Introduce `addr` argument to PCI device configs and update parser

Updates all config structs in order to make the new config option
available to all PCI device. Additionally update the parser so the new
option becomes available on the CLI.

Signed-off-by: Pascal Scholz <pascal.scholz@cyberus-technology.de>
On-behalf-of: SAP pascal.scholz@sap.com

amphi reviewed (approved):
Looks good to me!

phip1611 commented on vmm/src/device_manager.rs:
```diff
@@ -1631,7 +1619,7 @@ impl DeviceManager {
         let mut iommu_attached_devices = Vec::new();
         {
-            for handle in virtio_devices {
+            for handle in self.virtio_devices.clone() {
```
WAIT! This is an expensive clone, right? Why not just take everything by reference?

scholzp replied:
After reviewing the situation, we decided that we for now go with this solution. While the clone is still not nice, the new solution is still an improvement to the old one. We we'll see what improvements upstream suggests.

phip1611 replied:
For the record: If we change this, for example to

```patch
diff --git a/vmm/src/device_manager.rs b/vmm/src/device_manager.rs
index 4206aa155..953a6e24d 100644
--- a/vmm/src/device_manager.rs
+++ b/vmm/src/device_manager.rs
@@ -1620,7 +1620,7 @@ impl DeviceManager {
 
         let mut iommu_attached_devices = Vec::new();
         {
-            for handle in self.virtio_devices.clone() {
+            for handle in &self.virtio_devices {
                 let mapping: Option<Arc<IommuMapping>> = if handle.iommu {
                     self.iommu_mapping.clone()
                 } else {
@@ -1628,11 +1628,11 @@ impl DeviceManager {
                 };
 
                 let dev_id = self.add_virtio_pci_device(
-                    handle.virtio_device,
+                    handle.virtio_device.clone(),
                     &mapping,
-                    handle.id,
+                    handle.id.clone(),
                     handle.pci_segment,
-                    handle.dma_handler,
+                    handle.dma_handler.clone(),
                     handle.bdf_device,
                 )?;
 
```

then 

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
    --> vmm/src/device_manager.rs:1630:30
     |
1623 |               for handle in &self.virtio_devices {
     |                             --------------------
     |                             |
     |                             immutable borrow occurs here
     |                             immutable borrow later used here
...
1630 |                   let dev_id = self.add_virtio_pci_device(
     |  ______________________________^
1631 | |                     handle.virtio_device.clone(),
1632 | |                     &mapping,
1633 | |                     handle.id.clone(),
...    |
1636 | |                     handle.bdf_device,
1637 | |                 )?;
     | |_________________^ mutable borrow occurs here
```

phip1611 commented on option_parser/src/lib.rs:
```diff
@@ -167,6 +169,36 @@ impl OptionParser {
             .is_some()
     }
 
+    pub fn get_pci_device_function(
```
Please add brief documentation about the inputs and the output format. Especially the input format.

phip1611 reviewed (approved):
Amazing work! Just one final nit.
